### DISCUSSION TITLE
Audit trail - content storage

### DISCUSSION THREAD

**Discussion by PUXMKU** [OrchardCMS/OrchardCore#16442] [Ideas]
I was just creating custom audit trail extension, when I noticed that AuditTrail currently duplicitly stores Documents in json format. As I look at the code I do not see a problem to take version data of content items straight from Document table. I suppose it may be better to deduplicate this data, what do you think?

**Comment by Piedone** [comment]
I believe the idea there was that while you may completely remove content items (https://github.com/OrchardCMS/OrchardCore/issues/6056) or other data, the Audit Trail should remain an immutable store of changes.

**Reply by Piedone** [reply]
Also note that Audit Trail entries can be automatically trimmed in the background, check out the Audit Trail settings.

**Reply by PUXMKU** [reply]
Thanks for response. I guess it could be good to think of some solution to keep database free of duplicates. Right now audit trail relies on versions stored in Document table, even though it has its own copy. 

I would do either:
a/ removing versions from content table and keeping SoT in audit trail
b/ removing content data from audit trail and keeping SoT in documents table - in case someone would want to completely remove content item, it would behave as soft delete until audit trail trimming is done.

**Reply by Piedone** [reply]
We should probably make Audit Trail work with its own copy, then.
